### ISSUE TITLE
TIP-271: Vote for SR in Smart Contract

### ISSUE THREAD

**Issue by yanghang8612** [tronprotocol/tips#271]
```
tip: 271
title: Vote for SR in Smart Contract
author: yanghang8612@163.com
status: Draft
type: Standards Track
category : VM
created: 2021-06-01
```

## Simple Summary

To provide vote related operations in TVM.

## Abstract

Vote & WithdrawReward operations in system contract are introduced, smart contract can vote for witness and get reward from the system.

## Motivation

Common user can vote for activated witness to get block reward. However, none-privatekey accounts, like smart contracts, can not get benefit from this incentive mechanism. so this TIP provide instructions to implement the incentive functionalities.

## Implementation

Basically we need at least 2 Instructions:

- VOTE (vote for SR)
- WITHDRAWBALANCE (withdraw rewards to SR)

We might need several instructions, such as

- REWARDBALANCE (query vote reward)
- ISSRCANDIDATE (distinguish witness address type)
- VOTECOUNT(query vote count)

Make staking reward happened in smart contract can give DApp developers more motivation to build their DApps and increase their income. I highly recommend to implement this.

**Comment by Lredhdx** [comment]
Good ! I think query functions can be achieved by precompile contracts.

**Comment by otakuinny** [comment]
Great idea!!! Now all the ponzi ROI contract and scam contracts can use the TRX in their contracts to vote and get even richer while diluting the voting rewards for everyone else. Fantastic idea. Good job!!


Just because you can doesn't mean you should.

**Comment by Benson0224** [comment]
This issue will be closed as it is already implemented in https://github.com/tronprotocol/tips/blob/master/tip-271.md
and it is already included in https://github.com/tronprotocol/java-tron/releases/tag/GreatVoyage-v4.3.0
